uaclient/contract.py

```diff
@@ -153,25 +153,34 @@ def request_resource_machine_access(
         return resource_access
 
     def request_machine_token_update(
-        self, machine_token, contract_id, machine_id=None
-    ):
+        self,
+        machine_token: str,
+        contract_id: str,
+        machine_id: str = None,
+        detach: bool = False,
+    ) -> "Dict":
         """Request machine token refresh from contract server.
 
         @param machine_token: The machine token needed to talk to
             this contract service endpoint.
         @param contract_id: Unique contract id provided by contract service.
         @param machine_id: Optional unique system machine id. When absent,
             contents of /etc/machine-id will be used.
+        @param detach: Boolean set True if detaching this machine from the
+            active contract. Default is False.
 
         @return: Dict of the JSON response containing refreshed machine-token
         """
+        method = "DELETE" if detach else "POST"
         data = self._get_platform_data(machine_id)
         headers = self.headers()
         headers.update({"Authorization": "Bearer {}".format(machine_token)})
         url = API_V1_TMPL_CONTEXT_MACHINE_TOKEN_UPDATE.format(
             contract=contract_id, machine=data["machineId"]
         )
-        response, headers = self.request_url(url, headers=headers, data=data)
+        response, headers = self.request_url(
+            url, headers=headers, method=method, data=data
+        )
         if headers.get("expires"):
             response["expires"] = headers["expires"]
         self.cfg.write_cache("machine-token", response)
```

Comment:
This seems superfluous on detach; is there any reason to write this right before we delete it?